watcherx/changefeed.go

```diff
@@ -199,3 +203,75 @@ func WatchChangeFeed(ctx context.Context, cx *sqlx.DB, tableName string, c Event
 
 	return d, nil
 }
+
+// deduplicate sents events from `events` to the `deduplicated` channel, but
+// deduplicates events that are sent multiple times. This is necessary, because
+// the CochroachDB changefeed has a atleast-once guarantee for change events,
+// meaning that events could be sent multiple times.
+//
+// For deduplication, the last x `pastEvents` are considered.
+func deduplicate(in <-chan Event, out chan<- Event, pastEvents int) {
+	go func() {
+		defer close(out)
+		previous := newRingBuffer(pastEvents)
+
+		for {
+			e, ok := <-in
+			if !ok {
+				return
+			}
+			if previous.Contains(e) {
+				// Ignore event
+				continue
+			} else {
+				previous.Add(e)
+				out <- e
+			}
+		}
+	}()
+}
+
+// ringBuffer is a data structure for constant-time set membership (through
+// `Contains`) while maintaining constant memory usage by keeping at most
+// `capacity` elements.
+//
+// ringBuffer is not safe for concurrent use.
+type ringBuffer struct {
+	capacity int
+	seen     map[uint64]struct{} // map for efficient Contains().
+	keys     []uint64            // ring buffer so we can evict events on FIFO basis.
+	keyIdx   int                 // index of the next key to be added.
+	hash     maphash.Hash        // hash function for string keys.
+}
+
+func newRingBuffer(capacity int) *ringBuffer {
+	return &ringBuffer{
+		capacity: capacity,
+		seen:     make(map[uint64]struct{}, capacity),
+		keys:     make([]uint64, capacity, capacity),
+	}
+}
+
+func (r *ringBuffer) key(el fmt.Stringer) uint64 {
+	r.hash.Reset()
+	r.hash.WriteString(el.String())
+	return r.hash.Sum64()
+}
+
+func (r *ringBuffer) Contains(el fmt.Stringer) bool {
+	_, ok := r.seen[r.key(el)]
+	return ok
+}
+
+func (r *ringBuffer) Add(el fmt.Stringer) {
+	// Evict the oldest key.
+	if oldestKey := r.keys[r.keyIdx%r.capacity]; oldestKey != 0 {
```

Comment:
Can you please add a test for eviction?

Reply:
Test for eviction is in https://github.com/hperl/x/blob/92477b79aca33bec98145f6704e7507156ce58d3/watcherx/changefeed_test.go#L209.

Btw, I changed the implementation to use a hash function (`hash/maphash`) in the ring buffer, however I don't deal with collisions. Is this sensible / a trade worth taking?

Reply:
It's important that things stay consistent here, it's better to send an event twice than it is to lose an event due to this.

Reply:
Yeah I think it makes sense to use sha1 or something here

Reply:
Done!